url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/CLARION.java

```diff
@@ -0,0 +1,783 @@
+					}
+				}
+			}
+		}
+		return PerformedAction;
+	}
+	
+	/**
+	 * Generates the pre-input to use inside of CLARION for meta-cognition performed in the
+	 * early stages of perception.
+	 * @param NewSensoryInformation The new sensory information being perceived.
+	 * @return The pre-input for use within CLARION.
+	 */
+	private DimensionValueCollection generatePreInput (DimensionValueCollection NewSensoryInformation)
+	{
+		DimensionValueCollection preInput = NewSensoryInformation.clone();
+		if(ms != null)
+		{
+			preInput.putAll(ms.calculateAllDriveStrengths(NewSensoryInformation).toDimensionValueCollection());
+		}
+		return preInput;
+	}
+	
+	/**
+	 * Generates the current input to use inside of CLARION given the current sensory information,
+	 * the current goal, the episodic memory, the MS, and the items in working memory.
+	 * @param CurrentSensoryInformation The current sensory information being perceived.
+	 * @return The current input for use within CLARION.
+	 */
+	private DimensionValueCollection generateCurrentInput (DimensionValueCollection CurrentSensoryInformation)
+	{
+		DimensionValueCollection current = CurrentSensoryInformation.clone();
+		//Populate the current input with the current goal from the goal structure.
+		if(gs != null && gs.getCurrentGoal() != null)
+		{
+			for(Dimension d : gs.getCurrentGoal().values())
+			{
+				if(current.containsKey(d.getID()))
+				{
+					Dimension cd = current.get(d.getID());
+					for(Value v : d.values())
+					{
+						if(cd.containsKey(v.getID()))
+							cd.get(v.getID()).setActivation(v.getActivation());
+						else
+							cd.put(v.getID(), v.clone());
+					}
+				}
+				else
+					current.put(d.getID(), d.clone());
+			}
+		}
+		//Populate the current input with any chunks in working memory. For all chunks in
+		//working memory that contain sensory information, the activations of the 
+		//dimension-value pairs from those chunks will only replace the activations of the 
+		//corresponding dimension-value pairs in the current input if those activations are less
+		//than the normalized activations of the dimension-value pairs in working memory.
+		if(wm != null)
+		{
+			for(GenericChunkCollection gc : wm.get(WorkingMemory.InfoStored.FROM_NACS))
+			{
+				for(AbstractChunk c : gc.values())
+				{
+					for(Dimension d : c.values())
+					{
+						if(current.containsKey(d.getID()))
+						{
+							Dimension cd = current.get(d.getID());
+							for(Value v : d.values())
+							{
+								double act = v.getActivation() * InputSpace.get(d.getID()).get(v.getID()).getNormalizedBLA(Birthdate + InternalClock);
+								Value cv = cd.get(v.getID());
+								if(cv == null)
+								{
+									Value nv = v.clone();
+									nv.setActivation(act);
+									cd.put(nv.getID(),nv);
+								}
+								else if (cv.getActivation() < act)
+									cv.setActivation(act);
+							}
+						}
+						else
+							current.put(d.getID(), d.clone());
+					}
+				}
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 403:430, 431:458.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.